### ISSUE TITLE
Core/Creature: CREATURE_DIFFICULTYFLAGS_UNK10 is our new CREATURE_FLAG_EXTRA_PREVENT_AUTO_ATTACK

### ISSUE THREAD

**Issue by offl** [TrinityCore/TrinityCore#25677]
**Description:**

List of creatures with that flag:
```sql
SELECT * FROM creature_template WHERE type !=11 and entry IN (416,510,883,1419,1921,1933,2098,2151,2442,2523,2614,2615,2630,2637,2638,2665,2667,2671,2673,2674,2992,3444,3527,3560,3573,3579,3681,3844,3902,3903,3904,3906,3907,3908,3909,3911,3912,3913,3968,4277,4781,4952,4955,4957,4975,5187,5202,5625,5652,5674,5723,5781,5873,5874,5879,5913,5919,5920,5921,5922,5923,5924,5925,5926,5927,5929,5950,6012,6016,6017,6066,6110,6111,6112,6386,7366,7367,7368,7398,7399,7400,7402,7403,7412,7413,7414,7415,7416,7423,7424,7425,7464,7465,7466,7467,7468,7469,7483,7484,7486,7487,7527,7732,7785,7844,7845,8035,8179,8510,8658,8881,9496,9637,9658,9687,9688,9689,9699,9707,10183,10217,10218,10415,10467,10557,10685,10718,10780,11100,11101,11144,11179,11199,11446,11623,11711,11859,11875,12141,12276,12381,12382,12385,12426,12861,12922,13148,13221,13444,13445,13619,13838,13916,14081,14122,14361,14362,14386,14422,14434,14465,14466,14495,14662,14663,14664,14666,14751,14752,14801,14870,14987,14989,15047,15112,15141,15185,15224,15261,15276,15304,15328,15363,15364,15368,15427,15428,15430,15439,15447,15463,15464,15470,15474,15478,15479,15480,15482,15483,15484,15485,15486,15487,15488,15489,15490,15492,15496,15497,15514,15589,15712,15727,15769,15770,15771,15803,15804,15805,15957,16006,16111,16121,16129,16136,16140,16172,16182,16188,16211,16372,16385,16420,16431,16474,16479,16531,16598,16779,16887,16914,17001,17059,17060,17061,17096,17161,17265,17274,17283,17367,17368,17369,17378,17407,17408,17419,17458,17471,17503,17539,17578,17612,17652,17677,17832,17907,17973,17987,17990,18036,18095,18176,18177,18178,18179,18235,18236,18254,18370,18374,18412,18506,18691,18729,18795,18961,18967,18968,19139,19222,19224,19225,19230,19237,19327,19336,19381,19399,19403,19404,19405,19427,19430,19522,19636,19637,19638,19639,19665,19681,19750,19781,19782,19783,19869,19883,19922,20061,20143,20208,20214,20229,20243,20343,20411,20479,20605,20899,20979,21002,21011,21034,21091,21135,21185,21233,21322,21335,21364,21365,21369,21414,21420,21459,21654,21659,21703,21704,21705,21729,21731,21761,21812,21848,21876,21880,21926,21974,21992,21994,21995,22061,22069,22071,22078,22086,22091,22104,22124,22152,22161,22177,22189,22209,22228,22236,22250,22258,22279,22280,22315,22318,22328,22333,22335,22336,22418,22443,22444,22451,22461,22486,22487,22509,22868,22894,22895,22896,22897,22950,22970,22971,23041,23080,23082,23085,23095,23105,23123,23254,23279,23312,23313,23351,23366,23398,23412,23417,23426,23429,23438,23442,23451,23470,23491,23496,23499,23502,23700,23702,23706,23757,23771,23775,23777,23803,23805,23808,23811,23817,23822,23826,23867,23876,23884,23885,23915,23916,23917,23947,23957,23968,23972,24000,24008,24042,24045,24046,24108,24109,24165,24166,24167,24182,24183,24184,24185,24187,24193,24194,24210,24222,24224,24280,24284,24312,24320,24335,24336,24372,24373,24418,24447,24464,24640,24641,24649,24666,24675,24708,24745,24776,24792,24815,24861,24865,24879,24893,24912,24913,24921,24959,24961,24992,25040,25090,25091,25092,25219,25225,25265,25267,25297,25349,25384,25391,25422,25424,25444,25458,25502,25534,25583,25603,25741,25744,25765,25781,25855,25882,25960,25961,25987,26048,26086,26098,26403,26534,26788,26807,26808,26902,26925,26933,26937,27339,27392,27395,27430,27448,27460,27608,27609,27622,27629,27664,27692,27693,27697,27755,27756,27802,27829,27832,27838,27850,27878,27881,27883,27889,27894,27905,27910,27996,28058,28066,28094,28189,28230,28232,28239,28306,28308,28310,28311,28312,28319,28332,28366,28408,28413,28451,28452,28471,28563,28640,28664,28724,28781,28804,28833,28850,28854,28885,28887,28926,28938,29010,29105,29125,29126,29127,29134,29195,29224,29231,29271,29345,29368,29399,29401,29403,29458,29475,29483,29571,29684,29746,29747,29748,29752,29776,29789,29790,29791,29881,29929,29933,29998,30103,30134,30234,30248,30249,30268,30270,30273,30313,30330,30332,30391,30435,30491,30560,30563,30609,30616,30628,30641,30647,30652,30653,30654,30853,30879,30882,30887,30948,31075,31103,31119,31120,31121,31122,31129,31131,31132,31133,31144,31146,31157,31158,31162,31164,31165,31166,31167,31168,31169,31170,31171,31172,31173,31174,31175,31176,31181,31182,31185,31186,31189,31190,31204,31216,31280,31526,31553,31562,31576,31637,31638,31641,31644,31645,31646,31653,31688,31736,31765,31770,31782,31827,31830,31831,31857,31858,31861,31862,31881,32155,32186,32198,32208,32225,32246,32254,32282,32286,32388,32389,32406,32473,32504,32522,32535,32541,32542,32543,32544,32545,32546,32547,32567,32570,32574,32582,32627,32629,32633,32634,32635,32636,32637,32640,32666,32667,32775,32776,32789,32823,32826,32887,32898,32953,32955,33051,33052,33054,33059,33060,33061,33062,33063,33067,33080,33090,33108,33109,33114,33134,33139,33142,33143,33146,33167,33169,33173,33185,33189,33214,33217,33218,33229,33243,33248,33264,33272,33280,33288,33329,33364,33366,33369,33387,33433,33516,33531,33552,33567,33571,33575,33632,33651,33660,33661,33670,33705,33708,33716,33717,33718,33719,33720,33721,33725,33742,33750,33751,33752,33753,33782,33783,33802,33809,33835,33845,33863,33874,33881,33883,33890,33943,33983,33985,33990,33991,34055,34072,34099,34107,34120,34146,34150,34151,34159,34161,34181,34194,34202,34203,34213,34275,34286,34288,34307,34313,34428,34441,34444,34445,34447,34449,34450,34451,34459,34460,34465,34466,34468,34469,34470,34473,34474,34628,34630,34632,34658,34660,34686,34687,34704,34720,34755,34775,34776,34777,34778,34781,34793,34802,34812,34813,34819,34822,34823,34824,34825,34929,34935,34944,35014,35018,35062,35069,35226,35227,35228,35273,35335,35336,35376,35474,35493,35614,35633,35634,35635,35636,35637,35638,35640,35641,35644,35768,35819,35820,35821,36093,36215,36355,36356,36510,36532,36535,36542,36549,36550,36556,36557,36558,36559,36672,36838,36839,36847,36855,36968,36969,36978,36980,36982,36983,37041,37044,37074,37116,37117,37130,37215,37227,37488,37540,37543,37547,37574,37593,37749,37750,37751,37752,37766,37767,37768,37769,37770,37771,37772,37773,37810,37994,38008,38180,38332,38451,38454,38458,38566,38595,38667,38752,38903,38907,39259,39735,39743,39744,39841,40187,40281,40426,40725);
```
I ignored type totem to make result more readable. Unfortunatelly not all totems uses that flag, here's list of remaining:
```sql
SELECT * FROM creature_template WHERE entry IN (14667,17365,17366,18186,19897,19898,19899,19900,20455,20687,22322,23019,25455,25494,27527,28253,29225,29360,30444,31883,31893,31894,31895,31896,31897,32304,35311,36135,36605);
```
Note: if it contains heroic entry without normal entry, normal entry uses that flag.

Ok, more than 1k creatures, including critters, almost all totems and tons of temp summons and static creatures. 99.9% of them should never attack in melee. For sure that flag does not prevent creature from entering in melee range, it just prevents melee auto attack. So yeah, all creatures with that flag can chase player. Lots of creatures can chase player but can't attack in melee, mostly all of them uses damaging spells(tornados, lightning spheres, etc.)

The problem is it's not permanent flag, creature can remove it after specific event, usually it's spawn-in event. Luckily really small amount of creatures behaves like that.
I really don't understand why [Doomguard](https://tcubuntu.northeurope.cloudapp.azure.com/aowow/?npc=11859) uses that flag and I really hope he can't attack right after spawn but only after spawn-in event.
I checked multiple creatures but of course not all. For example Flame Leviathan's [Writhing Lasher](https://tcubuntu.northeurope.cloudapp.azure.com/aowow/?npc=33387) can attack in melee but he probably too can't attack right after spawn. Also there is [Dr. Boom](https://tcubuntu.northeurope.cloudapp.azure.com/aowow/?npc=20284) and he too can attack in melee. Maybe he uses [Dr. Boom Relay Invisibility](https://tcubuntu.northeurope.cloudapp.azure.com/aowow/?spell=35344) spell with Spawn effect. Maybe not.

It'll be nice if more people can check the list of creatures with that flag and cases like Doomguard and other.

**Expected behaviour:**

We need CREATURE_FLAG_EXTRA_PREVENT_AUTO_ATTACK and possibility to remove its effect in core scripts and SAI.
Prevent\allow auto attack is implemented in SAI with SMART_ACTION_AUTO_ATTACK and it will be good to be able to use it to remove effect of that flag.

**Branch(es):** both
**TC rev. hash/commit:** b52c89f1bae1

**Comment by offl** [comment]
Weirdly enough some creatures are rooted, can't melee but they're not passive. All this stuff is controlled by flags, sometimes creature has all 3 flags. But if creature is not passive, it can rotate in combat. For sure eggs should not rotate in combat and we have to change react state to passive in this case manually, without flag.

The thing is passive react state already roots creature and prevents melee so why don't just use it? One egg uses all 3 flags and another just is rooted and can't melee. Probably in some cases Blizz may disable rotate manually, probably it's handled somewhere else.

**Comment by illfated** [comment]
TrinityCore rev. b2fd442f64c5 2020-11-26 03:37:24 +0200 (3.3.5 branch) (Win64, Release, Static)

Not sure if it is directly related, but anyway: Funnily enough, some NPC barrels and other objects with gossip still rotate to "face" the player as long as the gossip is open. Example:

<details>
<summary>Quest interaction</summary>

Quest accepted: Can't Get Enough Turkey
Quest accepted: Don't Forget The Stuffing!
You gain 4350 experience.
Can't Get Enough Turkey completed.
Experience gained: 4350.
Received item: [Turkey Shooter].
You can only complete 24 more daily quests today.
You gain 4350 experience.
Don't Forget The Stuffing! completed.
Experience gained: 4350.
Received item: [Turkey Shooter].
You can only complete 23 more daily quests today.


</details>

```cpp
NPC currently selected by player:
Name: Bountiful Barrel
SpawnID: 138385, current GUID low: 446.
Entry: 35342.
Faction: 714.
NPC Flags: 640.
DisplayID: 29775 (Native: 29775).
Compatibility Mode: 0
Level: 40.
EquipmentId: 0 (Original: 0).
Health (base): 1524. (max): 1524. (current): 1524.
Movement type: Ground: Run, Swim: true, Flight: None, Chase: Run, Random: Walk
Unit Flags: 33536
* UNIT_FLAG_IMMUNE_TO_PC (0x100)
* UNIT_FLAG_IMMUNE_TO_NPC (0x200)
* UNIT_FLAG_SWIMMING (0x8000)
Unit Flags 2: 2048.
Dynamic Flags: 0.
Faction Template: 714.
SpawnTime: Full:5m Remaining:0s
Loot: 0 Pickpocket: 0 Skinning: 0
InstanceID: 0
Phasemask: 1
Armor: 1938
Position: 1293.329956 -4420.439941 26.589701.
AIName:  ScriptName: 
React state: AGGRESSIVE
Active AI: class AggressorAI
Flags Extra: 64
* CREATURE_FLAG_EXTRA_NO_XP_AT_KILL (0x40)
* is vendor (generic) (0x80)
* is vendor (food) (0x200)
MechanicImmuneMask: 0
```

```sql
11:03:22 Snurk Bucksquick yells: The zeppelin to Grom'gol has just arrived! All aboard for Stranglethorn!
```

**Comment by xvwyh** [comment]
@illfated, not related to this issue at all. It most likely needs to be fixed like in https://github.com/TrinityCore/TrinityCore/commit/b95c1520ab5d4464a793a862ed71ccee2d3163b1 - by adding a stun aura.

**Comment by offl** [comment]
> I really don't understand why Doomguard uses that flag and I really hope he can't attack right after spawn but only after spawn-in event.

Kinda solved

<details>

ServerToClient: SMSG_UPDATE_OBJECT (0x4715) Length: 426 ConnIdx: 2 Time: 05/31/2012 20:53:23.866 Number: 81516
[1] UpdateType: CreateObject2
[1] GUID: Full: 0xF1302E5300004019 Type: Creature Entry: 11859 Low: 16409

https://tcubuntu.northeurope.cloudapp.azure.com/aowow/?spell=42010

ServerToClient: SMSG_SPELL_START (0x6415) Length: 35 ConnIdx: 2 Time: 05/31/2012 20:53:23.866 Number: 81517
Caster GUID: Full: 0xF1302E5300004019 Type: Creature Entry: 11859 Low: 16409
Caster Unit GUID: Full: 0xF1302E5300004019 Type: Creature Entry: 11859 Low: 16409
Cast Count: 0
Spell ID: 42010 (42010)
Cast Flags: 2 (HasTrajectory)
Time: 0
Time2: 0
Target Flags: 0 (Self)

https://www.wowhead.com/spell=85692/doom-bolt

ServerToClient: SMSG_SPELL_START (0x6415) Length: 42 ConnIdx: 2 Time: 05/31/2012 20:53:26.269 Number: 81578
Caster GUID: Full: 0xF1302E5300004019 Type: Creature Entry: 11859 Low: 16409
Caster Unit GUID: Full: 0xF1302E5300004019 Type: Creature Entry: 11859 Low: 16409
Cast Count: 0
Spell ID: 85692 (85692)
Cast Flags: 2 (HasTrajectory)
Time: 0
Time2: 2585
Target Flags: 2 (Unit)
Target GUID: Full: 0xF13048EB00003EC9 Type: Creature Entry: 18667 Low: 16073

</details>

**Comment by offl** [comment]
> CREATURE_DIFFICULTYFLAGS_3_DISABLE_TURN                   = 0x02000000, // UNIT_FLAG2_DISABLE_TURN

is used by really small amount of creatures or is defined wrongly, it is used by all Pilgrim's Bounty vehicles
Anyway https://github.com/TrinityCore/TrinityCore/issues/25677#issuecomment-735023282 is something for another issue

**Comment by Rushor** [comment]
Sadly we cannot change flag extra values while Server is running

**Comment by offl** [comment]
We can, by adding CanAttackInMelee and set it to true or false if needed

We also can remove DoMeleeAttackIfReady, handle it in global AI and set CanAttackInMelee each time it's needed. It's how it's supposed to work and looks more logically correct than adding DoMeleeAttackIfReady each time in script
